framework/src/problems/DisplacedProblem.C

```diff
@@ -676,9 +676,10 @@ DisplacedProblem::addCachedResidual(THREAD_ID tid)
 void
 DisplacedProblem::addCachedResidualDirectly(NumericVector<Number> & residual, THREAD_ID tid)
 {
-
-  _assembly[tid]->addCachedResidual(residual, _displaced_nl.timeVectorTag());
-  _assembly[tid]->addCachedResidual(residual, _displaced_nl.nonTimeVectorTag());
+  if (_displaced_nl.hasVector(_displaced_nl.timeVectorTag()))
+    _assembly[tid]->addCachedResidual(residual, _displaced_nl.timeVectorTag());
+  if (_displaced_nl.hasVector(_displaced_nl.nonTimeVectorTag()))
+    _assembly[tid]->addCachedResidual(residual, _displaced_nl.nonTimeVectorTag());
 }
```

Comment:
I already check this in` Assembly.C`

```
void
Assembly::addCachedResidual(NumericVector<Number> & residual, TagID tag_id)
{
  if (!_sys.hasVector(tag_id))
  {
    _cached_residual_values[tag_id].clear();
    _cached_residual_rows[tag_id].clear();
    return;
  }
```

Does not this work for you? What is the issue here I am missing.

Reply:
Also, check `void
FEProblemBase::addCachedResidualDirectly(NumericVector<Number> & residual, THREAD_ID tid)`. There we do:
```
  if (_nl->hasVector(_nl->timeVectorTag()))
    _assembly[tid]->addCachedResidual(residual, _nl->timeVectorTag());
  if (_nl->hasVector(_nl->nonTimeVectorTag()))
    _assembly[tid]->addCachedResidual(residual, _nl->nonTimeVectorTag());
```

So I mimicked that in the `DisplacedProblem`.

Reply:
I see. I will check why Assembly does not take care of this.